### PR TITLE
Adds adaptive rate limiting using AIMD (Additive Increase, Multiplicative Decrease) backoff algorithm that dynamically adjusts based on downstream feedback signals.

### PR DESCRIPTION
**Pull request by swiftyspiffy**
## Summary
Adds adaptive rate limiting using AIMD (Additive Increase, Multiplicative Decrease) backoff algorithm that dynamically adjusts based on downstream feedback signals.

## Features
- **BackoffRateLimiter interface** for dynamic rate adjustment
- **RateLimitError type** to signal when rate limits are hit
- **AIMDRateLimiter** with configurable min/max bounds
- Automatic backoff on rate limit errors (×0.5)
- Automatic increase on success (+1 req/s)
- Thread-safe concurrent access
- Backward compatible with standard `rate.Limiter`

## Usage
```go
states := []jorb.State[AC, OC, JC]{
   {
       TriggerState: "api-call",
       Exec:         makeAPICall,
       Concurrency:  10,
       RateLimit:    jorb.NewAIMDRateLimiter(100, 10, 200), // start, min, max req/s
   },
}

func makeAPICall(ctx context.Context, ac AC, oc OC, jc JC) (JC, string, []jorb.KickRequest[JC], error) {
   resp, err := api.Call()
   if isRateLimited(err) {
       return jc, "api-call", nil, &jorb.RateLimitError{Err: err}
   }
   return jc, "done", nil, nil
}
```

## Testing
- 17 comprehensive tests covering all functionality
- Unit tests for core AIMD behavior
- Edge case and boundary condition tests
- Concurrency safety tests
- Processor integration tests
- **89.6% code coverage**

## Example
See `examples/aimd/main.go` for a complete working example.

## Backward Compatibility
Fully backward compatible - existing code using `rate.NewLimiter()` continues to work unchanged.

### CODE REVIEW

File: processor.go [hunk 1]
```diff
@@ -400,8 +403,23 @@ func (s *StateExec[AC, OC, JC]) Run() {
...
 			if err != nil {
 				j.StateErrors[priorState] = append(j.StateErrors[priorState], err.Error())
+				
+				// Handle rate limit errors with backoff
+				if IsRateLimitError(err) {
+					if backoff, ok := s.state.RateLimit.(BackoffRateLimiter); ok {
+						backoff.Backoff()
+						if aimd, ok := backoff.(*AIMDRateLimiter); ok {
+							slog.Info("Rate limit hit, backing off", "job", j.Id, "state", s.state.TriggerState, "newRate", aimd.Current())
+						}
```
Comment (gaffo) [inline processor.go]: Why do we need to test if it's a AIMD limiter here vs just putting this logging inside of the AIMDRateLimiter backoff function?

File: processor.go [hunk 2]
```diff
@@ -400,8 +403,23 @@ func (s *StateExec[AC, OC, JC]) Run() {
...
+					}
+				}
+				
 				slog.Info("Execution complete", "job", j.Id, "state", s.state.TriggerState, "newState", j.State, "error", err, "kickRequests", len(rtn.KickRequests))
 			} else {
+				// On success, increase rate if using AIMD
+				if aimd, ok := s.state.RateLimit.(*AIMDRateLimiter); ok {
+					aimd.onSuccess()
```
Comment (gaffo) [inline processor.go]: Same deal, why does this implementation in processor need to know it's an AIMD Rate Limiter instead of just a backoff? Shouldn't Backoff ratelimiter ust have a "success?

File: processor_test.go [hunk 3]
```diff
@@ -1009,3 +1009,208 @@ func TestProcessor_FirstStepExpands(t *testing.T) {
...
+
+	// Verify job completed
+	if r.Jobs["0"].State != "done" {
+		t.Errorf("Expected job to be in 'done' state, got '%s'", r.Jobs["0"].State)
+	}
+
+	// Verify errors were logged
+	if len(r.Jobs["0"].StateErrors["new"]) != 3 {
+		t.Errorf("Expected 3 errors logged for 'new' state, got %d", len(r.Jobs["0"].StateErrors["new"]))
+	}
```
[the remark this answers is not in the document]
Comment (gaffo) [reply processor_test.go]: Think about a test case where one person has 1000 tps rate limit and another has 1tps... for the 2nd person, this should ratchet down and find the 1s tps over time... how would we integration test that?

File: processor.go (hunk 1 above)
Comment (swiftyspiffy) [reply processor.go]: I think the thought was trying to decouple the rate limiter from specific logging behavior. That, in addition to the lack of context availability inside of `Backoff()`, so job id and trigger state not being available, though these aren't that important. 

Will update.

**Comment by gaffo** [comment]
Follow-up work has landed on `main` in this repo (injectable `Clock`, simplified AIMD config, `StateErrors` clone to fix race with serialization, README updates). See latest commits on `gaffo/jorb` `main`. This PR branch can be rebased/merged from `main` if you want to align the fork, or we can merge `main` after this PR closes.
